File: pruebatecnica-frontend/src/app/services/producto.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Producto } from '../models/producto';
import { environment } from '../../environments/environment';

@Injectable({ providedIn: 'root' })
export class ProductoService {
  private apiUrl = `${environment.apiUrl}/productos`;

  constructor(private http: HttpClient) {}

  listar(categoria?: string): Observable<Producto[]> {
    let params = new HttpParams();
    if (categoria) {
      params = params.set('categoria', categoria);
    }
    return this.http.get<Producto[]>(this.apiUrl, { params });
  }

  obtenerPorId(id: number): Observable<Producto> {
    return this.http.get<Producto>(`${this.apiUrl}/${id}`);
  }

  crear(producto: Producto): Observable<Producto> {
    return this.http.post<Producto>(this.apiUrl, producto);
  }

  actualizar(id: number, producto: Producto): Observable<Producto> {
    return this.http.put<Producto>(`${this.apiUrl}/${id}`, producto);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: pruebatecnicaSpringBoot/src/main/java/pruebatecnicaSpringBoot/pruebatecnicaSpringBoot/service/ProductoService.java
package pruebatecnicaSpringBoot.pruebatecnicaSpringBoot.service;

import pruebatecnicaSpringBoot.pruebatecnicaSpringBoot.dto.ProductoRequestDTO;
import pruebatecnicaSpringBoot.pruebatecnicaSpringBoot.dto.ProductoResponseDTO;

import java.util.List;

public interface ProductoService {
    List<ProductoResponseDTO> listar(String categoria);
    ProductoResponseDTO obtenerPorId(Long id);
    ProductoResponseDTO crear(ProductoRequestDTO productoRequestDTO);
    ProductoResponseDTO actualizar(Long id, ProductoRequestDTO productoRequestDTO);
    void eliminar(Long id);
}